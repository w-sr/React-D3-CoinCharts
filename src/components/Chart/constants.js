import { easeCubicOut } from "d3-ease";

// Chart
const CHART_PADDING_TOP = 20;
const TRANSITION = { duration: 500, ease: easeCubicOut };

// Cursor
const CURSOR_RADIUS_SIZE = 4;

// HorizontalAxis
const DEFAULT_TICK_COUNT = 7;

// HoverContainer
const HOVER_CONTAINER_WIDTH = 200;
const VERTICAL_OFFSET = -12;

export {
  CHART_PADDING_TOP,
  CURSOR_RADIUS_SIZE,
  DEFAULT_TICK_COUNT,
  HOVER_CONTAINER_WIDTH,
  TRANSITION,
  VERTICAL_OFFSET
};
