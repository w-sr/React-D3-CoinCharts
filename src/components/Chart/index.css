.chartContainer {
  cursor: crosshair;
  height: 221px;
  position: relative;
  width: 1060px;
}

.hoverContainer {
  display: flex;
  justify-content: center;
  position: absolute;
  width: 200px;
}

.hoverContainer.price {
  top: -12px;
}

.hoverContainer.time {
  bottom: -12px;
}

.hoverContent {
  border-radius: 5px;
  font-size: 14px;
  font-weight: 500;
  padding: 1px 6px;
}

.hoverContent.price {
  background: #7D95B6;
  color: #FFFFFF;
}

.hoverContent.time {
  background: #FFFFFF;
  border: 1px solid #7D95B6;
  color: #7D95B6;
}

svg {
  height: 100%;
  width: 100%;
}

.area {
  fill: #FFEBC5;
  pointer-events: none;
  stroke: none;
}

.line {
  fill: none;
  pointer-events: none;
  stroke: #FFB119;
  stroke-width: 2;
}

.cursorLine {
  stroke: #7D95B6;
  stroke-width: 1;
}

.activePoint{
  fill: white;
  stroke: #7D95B6;
  stroke-width: 2;
}