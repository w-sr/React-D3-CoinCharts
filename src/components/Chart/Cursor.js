import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import { animation, color, size } from "../../styles/constants";

const StyledCursor = styled.svg.attrs({
  style: ({ visible }) => ({
    opacity: visible ? 1 : 0,
  }),
})`
  height: 100%;
  width: 100%;
  transition: opacity ${animation.speed};
`;

const Circle = styled.circle`
  r: ${size.small};
  fill: ${color.white};
  stroke: ${color.coinchartsGray};
  stroke-width: 2;
`;

const Line = styled.line`
  stroke: ${color.coinchartsGray};
  stroke-width: 1;
`;

const Cursor = ({ height, visible, x, y }) => (
  <StyledCursor data-testid="Cursor" visible={visible}>
    <Line x1={x} x2={x} y1={0} y2={height} />
    <Circle cx={x} cy={y} />
  </StyledCursor>
);

Cursor.propTypes = {
  height: PropTypes.number.isRequired,
  visible: PropTypes.bool.isRequired,
  x: PropTypes.number.isRequired,
  y: PropTypes.number.isRequired,
};

export default Cursor;
