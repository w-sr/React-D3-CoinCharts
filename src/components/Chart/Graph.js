import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import isEqual from "lodash.isequal";
import { area as d3Area, line as d3Line } from "d3-shape";
import { extent } from "d3-array";
import { easeCubicOut } from "d3-ease";
import { interpolatePath } from "d3-interpolate-path";
import { scaleLinear, scaleTime } from "d3-scale";
import { select } from "d3-selection";
import "d3-transition";

import { PROPTYPES } from "../../constants";

export const GRAPH_PADDING_BOTTOM = 15; // spacing between lowest datum & border
export const GRAPH_PADDING_TOP = 15; // spacing between highest datum & border
const INITIAL_STATE = {
  previousColor: undefined,
  previousScaledData: [],
  scaledData: [],
  skipTransition: false,
};
const TRANSITION = { duration: 500, ease: easeCubicOut };

const StyledGraph = styled.svg`
  height: 100%;
  width: 100%;

  .area {
    pointer-events: none;
    stroke: none;
  }

  .line {
    fill: none;
    pointer-events: none;
    stroke-width: 2;
  }
`;

class Graph extends Component {
  static scaleData(data, height, width) {
    const scalePriceToY = scaleLinear()
      .range([height - GRAPH_PADDING_BOTTOM, GRAPH_PADDING_TOP])
      .domain(extent(data, d => d.price));

    const scaleTimeToX = scaleTime()
      .range([0, width])
      .domain(extent(data, d => d.time));

    return data.map(({ price, time }) => ({
      price: scalePriceToY(price),
      time: scaleTimeToX(time),
    }));
  }

  constructor(props) {
    super(props);
    this.state = INITIAL_STATE;
  }

  componentWillReceiveProps(nextProps) {
    const { data: nextData, height: nextHeight, width: nextWidth } = nextProps;
    const { color, width } = this.props;

    // Don't update if next set of data is not ready
    if (nextData.length === 0) {
      return;
    }

    this.setState(prevState => {
      const { scaledData } = prevState;
      const nextScaledData = Graph.scaleData(nextData, nextHeight, nextWidth);
      const previousScaledData =
        scaledData.length > 0 ? scaledData : nextScaledData.map(({ time }) => ({ price: nextHeight, time }));

      return {
        skipTransition: width !== nextWidth,
        previousColor: color,
        previousScaledData,
        scaledData: nextScaledData,
      };
    });
  }

  shouldComponentUpdate(nextProps) {
    return !isEqual(nextProps, this.props);
  }

  componentDidUpdate() {
    const { color, height } = this.props;
    const { previousColor = color, previousScaledData, scaledData, skipTransition } = this.state;
    const graph = select(this.svgNode);
    const transitionDuration = skipTransition ? 0 : TRANSITION.duration;

    const area = d3Area()
      .x(d => d.time)
      .y0(height)
      .y1(d => d.price);
    const line = d3Line()
      .x(d => d.time)
      .y(d => d.price);

    const previousAreaGraph = area(previousScaledData);
    const previousLineGraph = line(previousScaledData);
    const areaGraph = area(scaledData);
    const lineGraph = line(scaledData);

    graph.selectAll("path").remove();

    graph
      .append("path")
      .attr("class", "area")
      .attr("d", previousAreaGraph)
      .style("fill", previousColor.fill)
      .transition()
      .duration(transitionDuration)
      .ease(TRANSITION.ease)
      .attrTween("d", interpolatePath.bind(null, previousAreaGraph, areaGraph))
      .style("fill", color.fill);

    graph
      .append("path")
      .attr("class", "line")
      .attr("d", previousLineGraph)
      .style("stroke", previousColor.stroke)
      .transition()
      .duration(transitionDuration)
      .ease(TRANSITION.ease)
      .attrTween("d", interpolatePath.bind(null, previousLineGraph, lineGraph))
      .style("stroke", color.stroke);
  }

  render() {
    const svgRef = node => {
      this.svgNode = node;
    };
    return (
      <StyledGraph data-testid="Graph">
        <g ref={svgRef} />
      </StyledGraph>
    );
  }
}

Graph.propTypes = {
  color: PROPTYPES.COLOR.isRequired,
  data: PROPTYPES.PRICE_DATA.isRequired,
  height: PropTypes.number.isRequired,
  width: PropTypes.number.isRequired,
};

export default Graph;
