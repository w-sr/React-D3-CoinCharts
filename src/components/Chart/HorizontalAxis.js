import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import { extent } from "d3-array";
import { timeFormat } from "d3-time-format";
import styled from "styled-components";

import Flex from "../Flex";
import { DURATION, PROPTYPES } from "../../constants";
import { color, fontSize, fontWeight } from "../../styles/constants";

const StyledHorizontalAxis = styled(Flex).attrs({
  margin: props => (props.hideRightMargin ? "10px 0 0 20px" : "10px 20px 0 20px"),
})`
  margin: ${props => props.margin};
`;

const Tick = styled.div`
  color: ${color.coinchartsGray};
  font-size: ${fontSize.small};
  font-weight: ${fontWeight.medium};
`;

class HorizontalAxis extends PureComponent {
  static formatTime(timestamp, duration) {
    switch (duration) {
      case DURATION.ALL.key:
        return timeFormat("%b %Y")(timestamp); // 'Mmm YYYY'
      case DURATION.YEAR.key:
      case DURATION.MONTH.key:
      case DURATION.WEEK.key:
        return timeFormat("%b %_d")(timestamp); // 'Mmm D'
      case DURATION.DAY.key:
      case DURATION.HOUR.key:
      default:
        return timeFormat("%I:%M %p")(timestamp); // 'HH:MM PM/AM'
    }
  }

  static generateTicks(data, tickCount) {
    if (data.length < 2 || tickCount < 2) {
      return [];
    }

    const [minTime, maxTime] = extent(data, d => d.time);
    const rangeStep = (maxTime - minTime) / (tickCount - 1);
    const generatedTicks = [];
    for (let i = 0; i < tickCount; i += 1) {
      const time = new Date(minTime).valueOf();
      generatedTicks.push(time + i * rangeStep);
    }

    return generatedTicks;
  }

  render() {
    const { data, duration, tickCount, hideRightMargin } = this.props;
    const durationTicks = HorizontalAxis.generateTicks(data, tickCount);
    const axisTicks = durationTicks.map(timestamp => ({
      timestamp,
      label: HorizontalAxis.formatTime(timestamp, duration),
    }));

    return (
      <StyledHorizontalAxis data-testid="HorizontalAxis" justify="space-between" hideRightMargin={hideRightMargin}>
        {axisTicks && axisTicks.map(({ timestamp, label }) => <Tick key={timestamp}>{label}</Tick>)}
      </StyledHorizontalAxis>
    );
  }
}

HorizontalAxis.propTypes = {
  data: PROPTYPES.PRICE_DATA.isRequired,
  duration: PROPTYPES.DURATION.isRequired,
  tickCount: PropTypes.number.isRequired,
  hideRightMargin: PropTypes.bool.isRequired,
};

export default HorizontalAxis;
