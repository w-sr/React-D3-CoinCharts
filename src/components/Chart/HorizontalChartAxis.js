import React, { Component } from "react";
import PropTypes from "prop-types";
import isEqual from "lodash.isequal";
import { extent } from "d3-array";
import { timeFormat } from "d3-time-format";

import { DURATION, PROPTYPES } from "../../constants";
import { DEFAULT_TICK_COUNT } from "./constants";

class HorizontalChartAxis extends Component {
  static formatTime(timestamp, duration) {
    switch (duration) {
      case DURATION.ALL.key:
        return timeFormat("%b %Y")(timestamp); // 'Mmm YYYY'
      case DURATION.YEAR.key:
      case DURATION.MONTH.key:
      case DURATION.WEEK.key:
        return timeFormat("%b %_d")(timestamp); // 'Mmm D'
      case DURATION.DAY.key:
      case DURATION.HOUR.key:
      default:
        return timeFormat("%I:%M %p")(timestamp); // 'HH:MM PM/AM'
    }
  }

  static generateTimeAxisTicks(data, tickCount) {
    if (data.length < 2) {
      return [];
    }

    const [minTime, maxTime] = extent(data, d => d.time);
    const rangeStep = (maxTime - minTime) / (tickCount - 1);
    const generatedTicks = [];
    for (let i = 0; i < tickCount; i += 1) {
      const time = new Date(minTime).valueOf();
      generatedTicks.push(time + i * rangeStep);
    }

    return generatedTicks;
  }

  static renderTimeAxisTick(string, duration) {
    return (
      <div key={string} className="tick">
        <span>{HorizontalChartAxis.formatTime(string, duration)}</span>
      </div>
    );
  }

  // Only update when we receive new data
  shouldComponentUpdate(nextProps) {
    const { data } = this.props;
    const { data: nextData } = nextProps;

    // Don't update if next set of data is not ready
    if (nextData === undefined || nextData.length === 0) {
      return false;
    }

    return !isEqual(data, nextData);
  }

  render() {
    const { data, duration, tickCount } = this.props;
    const axisTicks = HorizontalChartAxis.generateTimeAxisTicks(
      data,
      tickCount
    );

    return (
      <div className="HorizontalChartAxis">
        {axisTicks &&
          axisTicks.map(time =>
            HorizontalChartAxis.renderTimeAxisTick(time, duration)
          )}
      </div>
    );
  }
}

HorizontalChartAxis.propTypes = {
  data: PROPTYPES.PRICE_DATA.isRequired,
  duration: PROPTYPES.DURATION.isRequired,
  tickCount: PropTypes.number
};

HorizontalChartAxis.defaultProps = {
  tickCount: DEFAULT_TICK_COUNT
};

export default HorizontalChartAxis;
