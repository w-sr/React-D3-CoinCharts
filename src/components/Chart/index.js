import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import isEqual from 'lodash.isequal';
import { area as d3Area, line as d3Line } from 'd3-shape';
import { extent } from 'd3-array';
import { interpolatePath } from 'd3-interpolate-path';
import { scaleLinear, scaleTime } from 'd3-scale';
import { select } from 'd3-selection';
import 'd3-transition';

import Cursor from './components/Cursor';

import {
  HOVER_CONTAINER_WIDTH,
  DEFAULT_COLOR,
  DEFAULT_TRANSITION,
  DEFAULT_FUNCTION,
} from './constants';
import { formatCurrency } from '../../utils';
import './index.css';

const ACTIVE_CURRENCY = 'usd';
const CHART_PADDING_TOP = 20;
const INITIAL_STATE = {
  data: [],
  dimensions: {
    height: 0,
    width: 0,
  },
  hovered: false,
  hoveredDataPoint: {},
  hoverX: -1,
  hoverY: -1,
  previousColor: DEFAULT_COLOR,
  previousScaledData: [],
  scaledData: [],
  scalePriceToY: DEFAULT_FUNCTION,
  scaleTimeToX: DEFAULT_FUNCTION,
};

class Chart extends Component {
  constructor(props) {
    super(props);
    this.state = INITIAL_STATE;
  }

  componentDidMount() {
    window.addEventListener('resize', this.handleResize);
    this.handleResize();
  }

  componentWillReceiveProps(nextProps) {
    const { data: nextData } = nextProps;
    const { color } = this.props;
    const { dimensions, scaledData } = this.state;

    const scaleTimeToX = scaleTime()
      .range([0, dimensions.width])
      .domain(extent(nextData, d => d.time));

    const scalePriceToY = scaleLinear()
      .range([dimensions.height, CHART_PADDING_TOP])
      .domain(extent(nextData, d => d.price));

    const nextScaledData = nextData.map(({ price, time }) => ({
      price: scalePriceToY(price),
      time: scaleTimeToX(time),
    }));

    const previousScaledData = (scaledData.length > 0) ?
      scaledData :
      nextScaledData.map(({ time }) => ({ price: dimensions.height, time }));

    this.setState({
      data: nextData,
      scaledData: nextScaledData,
      previousScaledData,
      previousColor: color,
      scaleTimeToX,
      scalePriceToY,
    });
  }

  // Only update when we receive new data or when the component is being hovered over
  shouldComponentUpdate(nextProps, nextState) {
    const { data } = this.props;
    const { dimensions, hovered, hoverX } = this.state;
    const { data: nextData } = nextProps;
    const {
      dimensions: nextDimensions,
      hovered: nextHovered,
      hoverX: nextHoverX,
    } = nextState;

    return (
      !isEqual(dimensions, nextDimensions) ||
      !isEqual(hovered, nextHovered) ||
      !isEqual(hoverX, nextHoverX) ||
      !isEqual(data, nextData)
    );
  }

  componentDidUpdate() {
    const { previousColor, previousScaledData, scaledData, dimensions } = this.state;
    const { color, transition } = this.props;
    const chart = select(this.svgNode);

    const area = d3Area()
      .x(d => d.time)
      .y0(dimensions.height)
      .y1(d => d.price);
    const line = d3Line()
      .x(d => d.time)
      .y(d => d.price);

    const previousAreaChart = area(previousScaledData);
    const previousLineChart = line(previousScaledData);
    const newAreaChart = area(scaledData);
    const newLineChart = line(scaledData);

    chart
      .selectAll('path')
      .remove();

    chart
      .append('path')
        .attr('class', 'Chart-area')
        .attr('d', previousAreaChart)
        .style('fill', previousColor.fill)
      .transition()
        .duration(transition.duration)
        .ease(transition.ease)
        .attrTween('d', () => interpolatePath(previousAreaChart, newAreaChart))
        .style('fill', color.fill);

    chart
      .append('path')
        .attr('class', 'Chart-line')
        .attr('d', previousLineChart)
        .style('stroke', previousColor.stoke)
      .transition()
        .duration(transition.duration)
        .ease(transition.ease)
        .attrTween('d', () => interpolatePath(previousLineChart, newLineChart))
        .style('stroke', color.stroke);

    // shouldComponentUpdate() ensures the following setState() doesn't cause a uneccesary rerender
    // eslint-disable-next-line react/no-did-update-set-state
    this.setState({ previousScaledData: scaledData });
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.handleResize);
  }

  handleResize = () => {
    const { data } = this.state;
    const { height, width } = this.chartSvgComponent.getBoundingClientRect();
    const dimensions = {
      height: Math.round(height),
      width: Math.round(width),
    };

    const scaleTimeToX = scaleTime()
      .range([0, dimensions.width])
      .domain(extent(data, d => d.time));

    const scalePriceToY = scaleLinear()
      .range([dimensions.height, CHART_PADDING_TOP])
      .domain(extent(data, d => d.price));

    const scaledData = data.map(({ price, time }) => ({
      price: scalePriceToY(price),
      time: scaleTimeToX(time),
    }));

    this.setState({
      data,
      dimensions,
      scaledData,
      previousScaledData: scaledData,
      scaleTimeToX,
      scalePriceToY,
    });
  }

  showHoverElements = () => {
    this.setState({ hovered: true });
  }

  hideHoverElements = () => {
    this.setState({ hovered: false });
  }

  updateHoverPosition = (e) => {
    const { data, dimensions, scalePriceToY } = this.state;
    const hoverX =
      e.nativeEvent.clientX - this.chartSvgComponent.getBoundingClientRect().left;

    // Find closest data point to the x-coordinates of where the user's mouse is hovering
    const index = Math.round((hoverX / dimensions.width) * (data.length - 1));
    const hoveredDataPoint = data[index] || {};

    const hoverY = scalePriceToY(hoveredDataPoint.price) || 0;

    this.setState({
      hoveredDataPoint,
      hoverX,
      hoverY,
      hovered: !!hoveredDataPoint,
    });
  };

  renderHoverElements() {
    const { hoveredDataPoint, hoverX, hovered } = this.state;
    const containerLeftPosition = hoverX - (HOVER_CONTAINER_WIDTH / 2);
    const displayClass = classNames({ 'Chart-show': hovered, 'Chart-hidden': !hovered });

    return (
      <div className={displayClass}>
        <div className="Chart-priceContainer" style={{ left: containerLeftPosition }}>
          <div className="content">{hoveredDataPoint.price && formatCurrency(hoveredDataPoint.price, ACTIVE_CURRENCY)}</div>
        </div>
        <div className="Chart-timeContainer" style={{ left: containerLeftPosition }}>
          <div className="content">{hoveredDataPoint.time && hoveredDataPoint.time.toLocaleString()}</div>
        </div>
      </div>
    );
  }

  render() {
    const { dimensions, hoverX, hoverY, hovered } = this.state;

    return (
      <div className="Chart-container">
        {this.renderHoverElements()}
        <svg
          ref={(node) => { this.chartSvgComponent = node; }}
          onMouseEnter={this.showHoverElements}
          onMouseLeave={this.hideHoverElements}
          onMouseMove={this.updateHoverPosition}
        >
          <g ref={(node) => { this.svgNode = node; }} />
          <Cursor
            height={dimensions.height}
            visible={hovered}
            x={hoverX}
            y={hoverY}
          />
        </svg>
      </div>
    );
  }
}

Chart.propTypes = {
  data: PropTypes.arrayOf(PropTypes.shape({
    price: PropTypes.number,
    time: PropTypes.data,
  })).isRequired,
  color: PropTypes.shape({
    fill: PropTypes.string,
    stroke: PropTypes.string,
  }),
  transition: PropTypes.shape({
    duration: PropTypes.number,
    ease: PropTypes.function,
  }),
};

Chart.defaultProps = {
  color: DEFAULT_COLOR,
  transition: DEFAULT_TRANSITION,
};

export default Chart;
