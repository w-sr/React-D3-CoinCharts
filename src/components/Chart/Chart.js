import React, { Component } from "react";
import PropTypes from "prop-types";
import { extent } from "d3-array";
import { scaleLinear } from "d3-scale";
import currencyFormatter from "currency-formatter";
import styled from "styled-components";

import { border } from "../../styles/constants";
import { PROPTYPES } from "../../constants";
import Flex from "../Flex";
import Cursor from "./Cursor";
import Graph, { GRAPH_PADDING_BOTTOM, GRAPH_PADDING_TOP } from "./Graph";
import HorizontalAxis from "./HorizontalAxis";
import HoverContainer from "./HoverContainer";
import VerticalAxis from "./VerticalAxis";

const DEFAULT_TICK_COUNT = 7;
const INITIAL_STATE = {
  dimensions: {
    height: 0,
    width: 0,
  },
  hovered: false,
  hoveredValue: {},
  hoverX: -1,
  hoverY: -1,
};

const StyledChart = styled.div`
  cursor: crosshair;
  height: 100%;
  width: 100%;
  position: relative;

  svg {
    position: absolute;
    left: 0;
    top: 0;
  }
`;

const StyledChartWithVerticalAxis = styled(Flex)`
  border-bottom: ${border.border};
  height: 225px;
`;

class Chart extends Component {
  constructor(props) {
    super(props);
    this.state = INITIAL_STATE;

    // Bind event-handlers
    this.handleResize = this.handleResize.bind(this);
    this.handleMouseEnter = this.handleMouseEnter.bind(this);
    this.handleMouseLeave = this.handleMouseLeave.bind(this);
    this.handleMouseMove = this.handleMouseMove.bind(this);
  }

  componentDidMount() {
    window.addEventListener("resize", this.handleResize);
    this.handleResize();
  }

  shouldComponentUpdate(nextProps) {
    const { isLoading } = nextProps;
    return !isLoading;
  }

  componentWillUnmount() {
    window.removeEventListener("resize", this.handleResize);
  }

  handleResize() {
    const { height, width } = this.chartSvgComponent.getBoundingClientRect();
    const dimensions = {
      height: Math.round(height),
      width: Math.round(width),
    };

    this.setState({ dimensions });
  }

  // Show hover elements
  handleMouseEnter() {
    this.setState({ hovered: true });
  }

  // Hide hover elements
  handleMouseLeave() {
    this.setState({ hovered: false });
  }

  // Update hover position
  handleMouseMove(e) {
    const { data, currency } = this.props;

    // Find closest data point to the x-coordinates of where the user's mouse is hovering
    const hoverX = e.nativeEvent.clientX - this.chartSvgComponent.getBoundingClientRect().left;

    this.setState(prevState => {
      const { dimensions } = prevState;
      const index = Math.round(hoverX / dimensions.width * (data.length - 1));
      const hoveredDatapoint = data[index] || {};
      const hoveredValue = {
        price: hoveredDatapoint.price && currencyFormatter.format(hoveredDatapoint.price, { code: currency }),
        time: hoveredDatapoint.time && hoveredDatapoint.time.toLocaleString(),
      };

      const scalePriceToY = scaleLinear()
        .range([dimensions.height - GRAPH_PADDING_BOTTOM, GRAPH_PADDING_TOP])
        .domain(extent(data, d => d.price));
      const hoverY = scalePriceToY(hoveredDatapoint.price) || 0;

      return {
        hovered: Boolean(hoveredDatapoint),
        hoveredValue,
        hoverX,
        hoverY,
      };
    });
  }

  render() {
    const {
      dimensions,
      hoveredValue: { price, time },
      hoverX,
      hoverY,
      hovered,
    } = this.state;
    const {
      color,
      currency,
      data,
      durationType,
      disableCursor,
      hideRightVerticalAxis,
      horizontalAxisTickCount,
    } = this.props;
    const svgRef = node => {
      this.chartSvgComponent = node;
    };

    return (
      <div>
        <StyledChartWithVerticalAxis>
          <VerticalAxis data={data} currency={currency} align="left" />
          <StyledChart
            data-testid="HoverRegion"
            innerRef={svgRef}
            onMouseEnter={this.handleMouseEnter}
            onMouseLeave={this.handleMouseLeave}
            onMouseMove={this.handleMouseMove}
          >
            <Graph color={color} data={data} height={dimensions.height} width={dimensions.width} />
            {!disableCursor && <Cursor height={dimensions.height} visible={hovered} x={hoverX} y={hoverY} />}
            {!disableCursor && <HoverContainer position="top" label={price} visible={hovered} x={hoverX} />}
            {!disableCursor && <HoverContainer position="bottom" label={time} visible={hovered} x={hoverX} />}
          </StyledChart>
          {!hideRightVerticalAxis && <VerticalAxis data={data} currency={currency} align="right" />}
        </StyledChartWithVerticalAxis>
        <HorizontalAxis
          data={data}
          duration={durationType}
          hideRightMargin={hideRightVerticalAxis}
          tickCount={horizontalAxisTickCount}
        />
      </div>
    );
  }
}

Chart.propTypes = {
  color: PROPTYPES.COLOR.isRequired,
  currency: PROPTYPES.CURRENCY.isRequired,
  data: PROPTYPES.PRICE_DATA.isRequired,
  durationType: PROPTYPES.DURATION.isRequired,
  isLoading: PropTypes.bool.isRequired,
  disableCursor: PropTypes.bool,
  hideRightVerticalAxis: PropTypes.bool,
  horizontalAxisTickCount: PropTypes.number,
};

Chart.defaultProps = {
  disableCursor: false,
  hideRightVerticalAxis: false,
  horizontalAxisTickCount: DEFAULT_TICK_COUNT,
};

export default Chart;
