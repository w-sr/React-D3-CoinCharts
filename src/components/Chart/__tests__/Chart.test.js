import React from "react";
import { render, Simulate } from "react-testing-library";

import Chart from "../";

const TEST_PROPS = {
  currency: "USD",
  data: [
    { price: 19.99, time: new Date(1000000) },
    { price: 29.99, time: new Date(2000000) },
    { price: 39.99, time: new Date(3000000) },
    { price: 49.99, time: new Date(4000000) },
  ],
  color: {
    fill: "#333",
    stroke: "#666",
  },
  durationType: "day",
  isLoading: false,
};

describe("<Chart/>", () => {
  it("renders without crashing", () => {
    const { container } = render(<Chart {...TEST_PROPS} />);
    expect(container.firstChild).not.toBeNull();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("does not rerender when `props.isLoading` is true", () => {
    const { container } = render(<Chart {...TEST_PROPS} />);
    const { textContent } = container.firstChild;

    // Component should not rerender
    render(<Chart {...TEST_PROPS} durationType="month" isLoading />, { container });
    expect(container.firstChild.textContent).toEqual(textContent);

    // Component should rerender
    render(<Chart {...TEST_PROPS} durationType="month" />, { container });
    expect(container.firstChild.textContent).not.toEqual(textContent);
  });

  it("renders child components correctly", () => {
    const { queryByTestId } = render(<Chart {...TEST_PROPS} />);
    expect(queryByTestId("Graph")).not.toBeNull();
    expect(queryByTestId("HorizontalAxis")).not.toBeNull();
    expect(queryByTestId("VerticalAxis")).not.toBeNull();
    expect(queryByTestId("Cursor")).not.toBeNull();
    expect(queryByTestId("HoverContainer")).not.toBeNull();
  });

  it("renders child components correctly when `props.disableCursor` is true", () => {
    const { queryByTestId } = render(<Chart {...TEST_PROPS} disableCursor />);
    expect(queryByTestId("Graph")).not.toBeNull();
    expect(queryByTestId("HorizontalAxis")).not.toBeNull();
    expect(queryByTestId("VerticalAxis")).not.toBeNull();
    expect(queryByTestId("Cursor")).toBeNull();
    expect(queryByTestId("HoverContainer")).toBeNull();
  });

  it("[DESKTOP] displays <HoverContainers/> when mouse is hovered over SVG chart", () => {
    const { getByTestId } = render(<Chart {...TEST_PROPS} />);

    // Simulate hovering (mouse entering chart)
    Simulate.mouseEnter(getByTestId("HoverRegion"));
    expect(getByTestId("HoverContainer").style.opacity).toEqual("1");

    // Simulate hovering (mouse leaving chart)
    Simulate.mouseLeave(getByTestId("HoverRegion"));
    expect(getByTestId("HoverContainer").style.opacity).toEqual("0");
  });
});
