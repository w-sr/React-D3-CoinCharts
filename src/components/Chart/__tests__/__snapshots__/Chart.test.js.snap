// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chart/> renders without crashing 1`] = `
<div>
  <div
    class="sc-gqjmRU kTfRCk sc-bdVaJa cAcgKX"
  >
    <div
      class="sc-dnqmqq eWTArw sc-bdVaJa ivjahH"
      data-testid="VerticalAxis"
    >
      <div
        class="sc-iwsKbI iGIseO sc-bdVaJa cAcgKX"
      >
        $50
      </div>
      <div
        class="sc-iwsKbI iGIseO sc-bdVaJa cAcgKX"
      >
        $20
      </div>
    </div>
    <div
      class="sc-gZMcBi jANtVc"
      data-testid="HoverRegion"
    >
      <svg
        class="sc-ifAKCX dCYFNS"
        data-testid="Graph"
      >
        <g />
      </svg>
      <svg
        style="opacity: 0;"
        class="sc-bwzfXH cZdfWA"
        data-testid="Cursor"
      >
        <line
          class="sc-bxivhb ftbvBS"
          x1="-1"
          x2="-1"
          y1="0"
          y2="0"
        />
        <circle
          class="sc-htpNat hKELQN"
          cx="-1"
          cy="-1"
        />
      </svg>
      <div
        class="sc-gzVnrw gCvPsj sc-bdVaJa epQZNM"
        style="top: -12px; left: -101px; opacity: 0;"
        data-testid="HoverContainer"
      >
        <div
          background="#7D95B6"
          border="none"
          color="#FFFFFF"
          class="sc-htoDjs hptRiM"
        />
      </div>
      <div
        class="sc-gzVnrw gCvPsj sc-bdVaJa epQZNM"
        style="bottom: -12px; left: -101px; opacity: 0;"
        data-testid="HoverContainer"
      >
        <div
          background="#FFFFFF"
          border="1px solid #DAE1E9"
          color="#7D95B6"
          class="sc-htoDjs RdJKU"
        />
      </div>
    </div>
    <div
      class="sc-dnqmqq fzzfzp sc-bdVaJa kAcxML"
      data-testid="VerticalAxis"
    >
      <div
        class="sc-iwsKbI iGIseO sc-bdVaJa cAcgKX"
      >
        $50
      </div>
      <div
        class="sc-iwsKbI iGIseO sc-bdVaJa cAcgKX"
      >
        $20
      </div>
    </div>
  </div>
  <div
    class="sc-EHOje hVdXPV sc-bdVaJa bJonzb"
    data-testid="HorizontalAxis"
  >
    <div
      class="sc-bZQynM itqFyr"
    >
      04:16 PM
    </div>
    <div
      class="sc-bZQynM itqFyr"
    >
      04:25 PM
    </div>
    <div
      class="sc-bZQynM itqFyr"
    >
      04:33 PM
    </div>
    <div
      class="sc-bZQynM itqFyr"
    >
      04:41 PM
    </div>
    <div
      class="sc-bZQynM itqFyr"
    >
      04:50 PM
    </div>
    <div
      class="sc-bZQynM itqFyr"
    >
      04:58 PM
    </div>
    <div
      class="sc-bZQynM itqFyr"
    >
      05:06 PM
    </div>
  </div>
</div>
`;
